File: src/clu/cmd/requires.py
import logging

from clu.config import get_config
from clu.input import check_file_exists, check_program_exists
from clu.opsys.factory import opsys_factory

log = logging.getLogger(__name__)
cfg = get_config()


def parse_args(subparsers):
    subp_requires = subparsers.add_parser(
        "requires", help="Check or list requirements for the system."
    )
    subp_requires.set_defaults(func=run)

    subp_requires.add_argument("subcmd", choices=["list", "check"], help="Sub-Command to run")


def run():
    log.info(f"Running command {cfg.cmd} with cfg={cfg}")
    if cfg.subcmd == "list":
        return list_requires()
    elif cfg.subcmd == "check":
        return check_requires()
    else:
        raise ValueError(f"Unknown sub-command: {cfg.subcmd}")


def list_requires() -> int:
    """List all the requirements for the current OS."""

    requires = opsys_factory().requires()

    print("Listing Requirements: ")
    print("----------------------")

    print("Files:")
    for file in requires.files:
        print(f"  - {file}")

    print("Programs:")
    for program in requires.programs:
        print(f"  - {program}")

    # print("APIs:")
    # for api in requires.apis:
    #     print(f"  - {api}")

    return 0


def check_requires() -> int:
    """Check if the current OS meets the requirements."""

    requires = opsys_factory().requires()
    print("Checking requirements:")
    print("----------------------")

    print("Files:")
    for file in requires.files:
        if check_file_exists(file):
            print(f"  - [ok] {file}")
        else:
            print(f"  - [MISSING] {file}")

    print("Programs:")
    for program in requires.programs:
        if check_program_exists(program):
            print(f"  - [ok] {program}")
        else:
            print(f"  - [MISSING] {program}")

    # print("APIs:")
    # for api in requires["apis"]:
    #     # what would go here?
    #     pass

    return 0

File: pkg/subcmd/requires.go
package subcmd

// Go port of src/clu/cmd/requires.py excluding parse_args. Provides listing and
// checking of system requirements gathered from the OpSys abstraction. Caller
// is expected to decide which subcommand (list or check) to invoke.

import (
	"fmt"
	"os"
	"strings"

	"github.com/NHGRI/clu/pkg/facts"
	"github.com/NHGRI/clu/pkg/input"
)

// RequiresCmd implements the "requires" subcommand (stub only).
type RequiresCmd struct {
	Mode string `arg:"" enum:"list,check" help:"Operation to perform: list or check."`
}

func (r *RequiresCmd) Run(stdout input.Stdout, stderr input.Stderr) error {
	switch r.Mode {
	case "list":
		// fmt.Println("requires: listing (stub)")
		listRequires(stdout, stderr)
	case "check":
		// fmt.Println("requires: checking (stub)")
		checkRequires(stdout, stderr)
	}
	return nil
}

// listRequires lists all file and program requirements.
func listRequires(stdout input.Stdout, stderr input.Stderr) int {
	reqs := facts.OpSysFactory().Requires()
	fmt.Fprintln(stdout, "Listing Requirements:")
	fmt.Fprintln(stdout, "----------------------")
	fmt.Fprintln(stdout, "Files:")
	for _, file := range reqs.Files {
		fmt.Fprintf(stdout, "  - %s\n", file)
	}
	fmt.Fprintln(stdout, "Programs:")
	for _, prog := range reqs.Programs {
		fmt.Fprintf(stdout, "  - %s\n", prog)
	}
	// APIs omitted; not used currently.
	return 0
}

// checkRequires checks existence of required files and programs.
func checkRequires(stdout input.Stdout, stderr input.Stderr) int {
	reqs := facts.OpSysFactory().Requires()
	fmt.Fprintln(stdout, "Checking requirements:")
	fmt.Fprintln(stdout, "----------------------")
	fmt.Fprintln(stdout, "Files:")
	for _, file := range reqs.Files {
		if checkFileExists(file) {
			fmt.Fprintf(stdout, "  - [ok] %s\n", file)
		} else {
			fmt.Fprintf(stdout, "  - [MISSING] %s\n", file)
		}
	}
	fmt.Fprintln(stdout, "Programs:")
	for _, prog := range reqs.Programs {
		if checkProgramExists(prog) {
			fmt.Fprintf(stdout, "  - [ok] %s\n", prog)
		} else {
			fmt.Fprintf(stdout, "  - [MISSING] %s\n", prog)
		}
	}
	return 0
}

// checkFileExists mimics Python check_file_exists helper using os.Stat.
func checkFileExists(path string) bool {
	if path == "" {
		return false
	}
	_, err := os.Stat(path)
	return err == nil
}

// checkProgramExists mimics Python check_program_exists by searching PATH.
func checkProgramExists(name string) bool {
	if name == "" {
		return false
	}
	// If the requirement includes spaces (arguments), use first token for lookup.
	prog := name
	for i, r := range name {
		if r == ' ' {
			prog = name[:i]
			break
		}
	}
	// Look through PATH entries.
	pathEnv := os.Getenv("PATH")
	if pathEnv == "" {
		return false
	}
	for _, dir := range splitPath(pathEnv) {
		candidate := dir + string(os.PathSeparator) + prog
		if fi, err := os.Stat(candidate); err == nil && !fi.IsDir() {
			return true
		}
	}
	return false
}

// splitPath splits PATH variable; separated for testability.
func splitPath(p string) []string { return strings.Split(p, string(os.PathListSeparator)) }
